### DISCUSSION TITLE
mounting cloud-init seed.iso

### DISCUSSION THREAD

**Discussion by dmgoldstein1** [utmapp/UTM#6201] [UTM Help (macOS)]
I'm trying to run Amazon Linux 2 on UTM on my MacBook Pro (M1 Pro Max, Ventura 13.6.2, UTM 4.4.5) AWS Linux 2 credentials are added to a new machine using cloud-init. I have followed instructions from Amazon, Canonical and others to try to properly create my user-data file, save it as seed.iso and have those credentials be the ones the VM uses when it is first booted, and so far it’s not working. I’ve even tried using a bare-bones template with a plain text password, just so I could troubleshoot the way I’m using cloud-init and UTM. 

I right-click on the VM on the left side of the main UTM window, scroll down to Drives, and add the .qcow2 file and the seed.iso file. I’ve tried various combinations of “Image Type” and “Interface” options.

I can launch Amazon Linux 2 from the .qcow2 file provided at https://cdn.amazonlinux.com/os-images/2.0.20240223.0/ but no matter what I’ve tried over many hours and days, the Linux VM boots up in the virtual UTM display, but the password and login attempts come back as “Login incorrect”. 

Does anyone have experience using cloud-init with UTM? Maybe I’m doing something wrong with the UTM GUI when I add the .iso file.

**Comment by DatDucati** [comment]
I have the same issue, but with Debian bullseye. The cloud init configuration is valid, because it does configure a vm on my proxmox server just fine. Any hints on how to debug this issue would be appreciated.

**Comment by quornicus** [comment]
TL;DR: try using a virtio drive, readonly and nonremovable, and importing the ISO. Make sure to set the 'Image Type' for the device to 'CD/DVD (ISO) Image'.

I experienced the same issue with Ubuntu 20.04, 22.04, and 24.04. I was able to get cloud-init to see my seed iso by:

- using a virtio drive, non-removable, read-only.
- 'importing' the iso and making sure I had selected 'CD/DVD (ISO) Image' from the 'Image Type' dropdown.

The only things remotely helpful for debugging was confirmation that my seed iso indeed worked by launching it with plain-old cli qemu. I also removed the 'Display' device and instead attached a serial device, which allowed me to scroll back and see the output of cfn-init after it ran. 

When using virtio for the seed image, the ISO is apparently converted to the qcow2 format, but the process works. 

Edit: This is on a MacBook Air M2, using Ubuntu LTS cloud images (20, 22, 24), UTM v 4.4.5.

**Comment by nileshtrivedi** [comment]
I am facing this issue as well (with Amazon Linux 2023). I removed the Display device and attached a serial device, but no matter what I do with the `seed.iso` disk configuration, cloud-init never shows up in the booting logs.

**Comment by nileshtrivedi** [comment]
I got it working with this config:

In `meta-data`:
```
hostname: myhost
create_hostname_file: true
```

In `user-data`:
```
#cloud-config
#vim:syntax=yaml
password: test123
chpasswd:
  expire: False
```

AL2023 ships with cloud-init 22.2.2 so make sure to read the right version of docs. Also, cloud-init only runs on the first boot, so every time you make a change to seed.iso, make sure to remove both the drives and reattach them.

**Comment by rinsuki** [comment]
Alternatively, if your VM has a NIC, you can use the nocloud-net data source by adding a QEMU argument.

![image](https://github.com/user-attachments/assets/4b83b1fe-68c5-45a5-a3ec-59023c9b985a)

**Reply by naveenrajm7** [reply]
Thank you!
This saves me from the extra step of building ISO files. 
I was trying `ds=nocloud` as per [cloud-init docs](https://cloudinit.readthedocs.io/en/latest/tutorial/qemu.html#launch-a-vm-with-our-user-data), it didn't work. nocloud-net did!

**Comment by patrostkowski** [comment]
I found even easier way to use cloud-init in UTM by following [this guide](https://codeofconnor.com/booting-cloud-images-with-qemu/).

```
$ touch meta-data
$ cat > user-data << EOF
#cloud-config

system_info:
   default_user:
     name: fedora

chpasswd:
  list: |
    fedora:password
  expire: False
 

resize_rootfs: True
ssh_authorized_keys:
   - ssh-rsa AAAAB3Nza...
EOF
$ genisoimage -output seedci.iso -volid cidata -joliet -rock user-data meta-data
```
When `seedci.iso` file is ready, all you have to do is update QEMU Arguments via UTM app by adding `-cdrom $PATH_TO_ISO`:

![image](https://github.com/user-attachments/assets/749cfb7f-3b0e-4d7a-9ecd-0e0901a64585)

Voilà! Cloud-init will be triggered after booting up the machine.

**Comment by problame** [comment]
@patrostkowski 's solution (manually add `-cdrom`) works for me with qemu https://github.com/utmapp/UTM/discussions/6201#discussioncomment-11439070

But how to make it work with Apple Virtualization?

**Reply by naveenrajm7** [reply]
Hi @problame 

This is what I did to use a cloud image and an init file with Apple virtualization. 

From https://cloud-images.ubuntu.com/releases/noble/release/unpacked/ got
* kernel image 
* ramdisk
used boot arguments : `console=hvc0 root=/dev/vda ds=nocloud;s=http://192.168.76.1:8080/ubuntu-cloud/ cloud-init.verbose=1`

From https://cloud-images.ubuntu.com/releases/noble/release/ got
* disk image
Attached this as the only disk

I had a simple Python web server hosting cloud-init files at http://locahost:8080/ubuntu-cloud/

![Screenshot 2025-03-28 at 8 41 53 PM](https://github.com/user-attachments/assets/3a6873ee-e6c5-4cb8-90bb-8d61000cbe9f)

**Reply by problame** [reply]
Hm, yeah, that works.

wo dering why it’s not possible to simply attach the ISO through the UI as removable read-only media though.

It appears that it **must** be the `-cdrom` option, and that obvsly only exists for qemu

**Reply by Apurvag011** [reply]
Hey can you give detailed steps please trying to create local k8s cluster with Ubuntu 22.04 and using cloud init for basic setup of binaries and setting up kernel parameters
